### PR TITLE
Add missing metaclasses

### PR DESCRIPTION
**Pull request by sobolevn**
Refs: https://github.com/python/mypy/pull/13331

### CODE REVIEW

File: stdlib/dataclasses.pyi [hunk 1]
```diff
@@ -217,7 +217,13 @@ def is_dataclass(obj: Any) -> bool: ...
 
 class FrozenInstanceError(AttributeError): ...
 
-class InitVar(Generic[_T]):
+if sys.version_info >= (3, 9):
+    _InitVarMeta = type
```
Comment (AlexWaygood) [inline stdlib/dataclasses.pyi:221]:
```suggestion
    _InitVarMeta: TypeAlias = type
```

File: stdlib/string.pyi [hunk 2]
```diff
@@ -32,7 +32,14 @@ whitespace: LiteralString
 
 def capwords(s: StrOrLiteralStr, sep: StrOrLiteralStr | None = ...) -> StrOrLiteralStr: ...
 
-class Template:
+if sys.version_info >= (3, 9):
+    _TemplateMetaclass = type
```
Comment (AlexWaygood) [inline stdlib/string.pyi:36]:
```suggestion
    _TemplateMetaclass: TypeAlias = type
```

File: stdlib/contextlib.pyi [hunk 3]
```diff
@@ -133,7 +134,8 @@ class _RedirectStream(AbstractContextManager[_T_io]):
 class redirect_stdout(_RedirectStream[_T_io]): ...
 class redirect_stderr(_RedirectStream[_T_io]): ...
 
-class ExitStack:
+# In reallity this is a subclass of `AbstractContextManager`:
```
Comment (AlexWaygood) [inline stdlib/contextlib.pyi:137]:
```suggestion
# In reality this is a subclass of `AbstractContextManager`;
# see #7961 for why we don't do that in the stub
```

File: stdlib/contextlib.pyi [hunk 4]
```diff
@@ -148,7 +150,8 @@ class ExitStack:
 _ExitCoroFunc: TypeAlias = Callable[[type[BaseException] | None, BaseException | None, TracebackType | None], Awaitable[bool]]
 _ACM_EF = TypeVar("_ACM_EF", bound=AbstractAsyncContextManager[Any] | _ExitCoroFunc)
 
-class AsyncExitStack:
+# In reallity this is a subclass of `AbstractAsyncContextManager`:
```
Comment (AlexWaygood) [inline stdlib/contextlib.pyi:153]:
```suggestion
# In reality this is a subclass of `AbstractAsyncContextManager`;
# see #7961 for why we don't do that in the stub
```

**Comment by sobolevn** [comment]
Ok, looks like mypy cannot work with `TypeAlias`ed metaclasses. I will send a fix shortly.

File: stdlib/dataclasses.pyi (hunk 1 above)
Comment (sobolevn) [reply stdlib/dataclasses.pyi:221]: Sadly, both do not work 😢 
See https://github.com/python/mypy/issues/13334

Comment (AlexWaygood) [reply stdlib/dataclasses.pyi:221]: :(

For now, you could probably put the whole class under a `sys.version_info` branch? The `__class_getitem__` method was added in 3.9 as well

**Comment by sobolevn** [comment]
This mypy PR should fix that: https://github.com/python/mypy/pull/13335

**Comment by srittau** [comment]
I marked this as deferred. As I understand it, we need to wait for the next mypy release to be able to land this.

**Comment by AlexWaygood** [comment]
> I marked this as deferred. As I understand it, we need to wait for the next mypy release to be able to land this.

We can do the ones in `ast.pyi` and `contextlib.pyi` now. @sobolevn, why don't you revert the other changes for now, and we can do them after the next mypy release?

**Comment by sobolevn** [comment]
Done! Good idea 👍

File: stdlib/ast.pyi [hunk 5]
```diff
@@ -5,21 +5,25 @@ from typing import Any, TypeVar, overload
 from typing_extensions import Literal
 
 if sys.version_info >= (3, 8):
-    class Num(Constant):
+    class _ABC(type):
+        if sys.version_info >= (3, 9):
+            def __init__(cls, *args: Any) -> None: ...
```
Comment (AlexWaygood) [inline stdlib/ast.pyi:10]: The `*args` seem to just be ignored at runtime.

```suggestion
            def __init__(cls, *args: object) -> None: ...
```

File: stdlib/contextlib.pyi [hunk 6]
```diff
@@ -133,7 +134,8 @@ class _RedirectStream(AbstractContextManager[_T_io]):
 class redirect_stdout(_RedirectStream[_T_io]): ...
 class redirect_stderr(_RedirectStream[_T_io]): ...
 
-class ExitStack:
+# In reality this is a subclass of `AbstractContextManager`:
```
Comment (AlexWaygood) [inline stdlib/contextlib.pyi:137]:
```suggestion
# In reality this is a subclass of `AbstractContextManager`;
# see #7961 for why we don't do that in the stub
```

File: stdlib/contextlib.pyi [hunk 7]
```diff
@@ -148,7 +150,8 @@ class ExitStack:
 _ExitCoroFunc: TypeAlias = Callable[[type[BaseException] | None, BaseException | None, TracebackType | None], Awaitable[bool]]
 _ACM_EF = TypeVar("_ACM_EF", bound=AbstractAsyncContextManager[Any] | _ExitCoroFunc)
 
-class AsyncExitStack:
+# In reality this is a subclass of `AbstractAsyncContextManager`:
```
Comment (AlexWaygood) [inline stdlib/contextlib.pyi:153]:
```suggestion
# In reality this is a subclass of `AbstractAsyncContextManager`;
# see #7961 for why we don't do that in the stub
```

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Comment by AlexWaygood** [comment]
@sobolevn, now that mypy works with `TypeAlias`'d metaclasses, do you want to do a new PR with the `dataclasses`/`string` changes?

**Comment by sobolevn** [comment]
Thanks for the reminder :) 
Will do in 1 hour.

**Pull request merged**